File: 2024/AoC24-02ab.swift
extension Sequence {
  func tupleWindows() -> [(Element, Element)] {
    return Array(zip(self, self.dropFirst()))
  }
}

struct Report {
  let values: [Int]

  func isSafe(tolerance: Int = 0) -> Bool {
    let sign = values[0] < values[1] ? 1 : -1
    var check = 0
    for (a, b) in values.map( { $0 * sign } ).tupleWindows() {
      if !(1...3).contains(b - a) {
        check += 1
        if check > tolerance { return false }
      }
    }

    return true
  }
}

func getInput() -> [Report] {
  var reports: [Report] = []
  while let line = readLine() {
    let nums = line.split(separator: " ").map { Int($0)! }
    reports.append(Report(values: nums))
  }

  return reports
}

let reports = getInput()

print("Part 1 answer:", reports.filter { $0.isSafe() }.count)
print("Part 2 answer:", reports.filter { $0.isSafe(tolerance: 1) }.count)

File: 2024/AoC24-02ab.rs
use std::fs;
use itertools::Itertools;
use std::ops::Sub;

fn main() {
  let input = fs::read_to_string("input.txt").unwrap();

  let reports: Vec<Report> = input.lines().map(process).collect();

  println!("Part 1 answer: {}", reports.iter().filter(|r| r.is_safe(0)).count());
  println!("Part 2 answer: {}", reports.iter().filter(|r| r.is_safe(1)).count());
}

fn process(line: &str) -> Report {
  Report {
    values: line.split(' ').map(|s| s.parse().unwrap()).collect(),
  }
}

#[derive(Debug, Clone)]
struct Report {
  values: Vec<i32>,
}

impl Report {
  fn is_safe(&self, tolerance: u32) -> bool {
    let sign =  if self.values[0] < self.values[1] { 1 } else { -1 };

    let mut check = 0;
    for (a, b) in self.values.iter().map(|n| n * sign).tuple_windows() {
      if !(1..=3).contains(&b.sub(a)) {
        check += 1;
        if check > tolerance { return false; }
      }
    }

    true
  }
}